packages/flutter_tools/lib/src/cache.dart

```diff
@@ -511,7 +511,16 @@ abstract class CachedArtifact extends ArtifactSet {
       }
     }
     await updateInner();
-    cache.setStampFor(stampName, version);
+    try {
+      cache.setStampFor(stampName, version);
+    } on FileSystemException catch (err) {
+      globals.printError(
+        'The new artifact "$name" was downloaded, but Flutter failed to update '
+        'its stamp file, receiving the error "$err". '
+        'Flutter can continue, but the artifact may be re-downloaded on '
+        'subsequent invocations until the problem is resolved.',
```

Comment:
Though this is an improvement, I'm not sure if a user would be able to figure out the problem here. If we could figure out that it is a permission issue and give a more specific message that would be great

Reply:
Yeah, I thought about that... Unfortunately, the error code in the `FileSystemException` will be different across OSes, and the error string will be different across OSes and locales. A Dart library to abstract away those things would be great, but doesn't exist.